kaldb/src/main/java/com/slack/kaldb/bulkIngestApi/BulkIngestKafkaProducer.java

```diff
@@ -0,0 +1,298 @@
+
+    checkArgument(
+        !kafkaConfig.getKafkaBootStrapServers().isEmpty(),
+        "Kafka bootstrapServers must be provided");
+
+    checkArgument(!kafkaConfig.getKafkaTopic().isEmpty(), "Kafka topic must be provided");
+
+    this.datasetMetadataStore = datasetMetadataStore;
+    this.pendingRequests = new LinkedBlockingQueue<>();
+
+    // todo - consider making this a configurable value or removing the config
+    this.producerSleep =
+        Integer.parseInt(System.getProperty("kalDb.bulkIngest.producerSleep", "50"));
+
+    // since we use a new transaction ID every time we start a preprocessor there can be some zombie
+    // transactions?
+    // I think they will remain in kafka till they expire. They should never be readable if the
+    // consumer sets isolation.level as "read_committed"
+    // see "zombie fencing" https://www.confluent.io/blog/transactions-apache-kafka/
+    this.kafkaProducer = createKafkaTransactionProducer(UUID.randomUUID().toString());
+    this.kafkaMetrics = new KafkaClientMetrics(kafkaProducer);
+    this.kafkaMetrics.bindTo(meterRegistry);
+
+    this.failedSetResponseCounter = meterRegistry.counter(FAILED_SET_RESPONSE_COUNTER);
+    this.stallCounter = meterRegistry.counter(STALL_COUNTER);
+    this.batchSizeGauge = meterRegistry.gauge(BATCH_SIZE_GAUGE, new AtomicInteger(0));
+
+    this.kafkaProducer.initTransactions();
+  }
+
+  private void cacheSortedDataset() {
+    // we sort the datasets to rank from which dataset do we start matching candidate service names
+    // in the future we can change the ordering from sort to something else
+    this.throughputSortedDatasets =
+        datasetMetadataStore.listSync().stream()
+            .sorted(Comparator.comparingLong(DatasetMetadata::getThroughputBytes).reversed())
+            .toList();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    cacheSortedDataset();
+    datasetMetadataStore.addListener(datasetListener);
+  }
+
+  @Override
+  protected void run() throws Exception {
+    while (isRunning()) {
+      List<BulkIngestRequest> requests = new ArrayList<>();
+      pendingRequests.drainTo(requests);
+      batchSizeGauge.set(requests.size());
+      if (requests.isEmpty()) {
+        try {
+          stallCounter.increment();
+          Thread.sleep(producerSleep);
+        } catch (InterruptedException e) {
+          return;
+        }
+      } else {
+        transactionCommit(requests);
+      }
+    }
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    datasetMetadataStore.removeListener(datasetListener);
+
+    kafkaProducer.close();
+    if (kafkaMetrics != null) {
+      kafkaMetrics.close();
+    }
+  }
+
+  public BulkIngestRequest submitRequest(Map<String, List<Trace.Span>> inputDocs) {
+    BulkIngestRequest request = new BulkIngestRequest(inputDocs);
+    pendingRequests.add(request);
+    return request;
+  }
+
+  protected Map<BulkIngestRequest, BulkIngestResponse> transactionCommit(
```

Comment:
maybe a better name could be `produceDocsAndCommit` ?

Reply:
Done - `produceDocumentsAndCommit`